## UAP 1 (Backpropagation)

Nama : Salma Adzra Fathina

NIM : 225150207111042

### Pelatihan Model dengan Stochastic Gradient Descent

Anda sedang mencoba untuk mengoptimalkan performa arsitektur JST anda yang dilatih menggunakan Backpropagation. Anda menyadari permasalahan bahwa model anda sering mengalami kegagalan mencapai konvergensi optimal dikarenakan adanya local minima. Untuk menyelesaikan permasalahan tersebut, anda menggunakan strategi pembaruan bobot yang lebih baru dengan menggunakan Stochastic Gradient Descent (SGD)

Catatan :
- Data memiliki campuran antara nilai kategorikal dengan nilai numerik, sedangkan Perceptron hanya mampu menggunakan nilai numerik
- Untuk menentukan apakah data berjenis kategorikal atau numerik dapat ditentukan dengan inspeksi data dan dokumentasi data yang diberikan
- Template kode yang diberikan dapat diubah sesuai dengan kebutuhan
- Perbedaan GD dengan SGD hanya pada proses update bobot yang dilakukan pada setiap data secara langsung

Peraturan :
- Open Modul
- Dilarang menggunakan AI (Chatbot, Code Generator, dll) untuk membantu pengerjaan
- Dilarang menggunakan library yang mengimplementasikan SOM siap pakai
- Terindikasi menggunakan AI ataupun library akan terkena penalti nilai

## Data Loading and Preprocessing

In [ ]:
from google.colab import files

files.upload()

Saving heart.csv to heart.csv


{'heart.csv': b'age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output\r\n63,1,3,145,233,1,0,150,0,2.3,0,0,1,1\r\n37,1,2,130,250,0,1,187,0,3.5,0,0,2,1\r\n41,0,1,130,204,0,0,172,0,1.4,2,0,2,1\r\n56,1,1,120,236,0,1,178,0,0.8,2,0,2,1\r\n57,0,0,120,354,0,1,163,1,0.6,2,0,2,1\r\n57,1,0,140,192,0,1,148,0,0.4,1,0,1,1\r\n56,0,1,140,294,0,0,153,0,1.3,1,0,2,1\r\n44,1,1,120,263,0,1,173,0,0,2,0,3,1\r\n52,1,2,172,199,1,1,162,0,0.5,2,0,3,1\r\n57,1,2,150,168,0,1,174,0,1.6,2,0,2,1\r\n54,1,0,140,239,0,1,160,0,1.2,2,0,2,1\r\n48,0,2,130,275,0,1,139,0,0.2,2,0,2,1\r\n49,1,1,130,266,0,1,171,0,0.6,2,0,2,1\r\n64,1,3,110,211,0,0,144,1,1.8,1,0,2,1\r\n58,0,3,150,283,1,0,162,0,1,2,0,2,1\r\n50,0,2,120,219,0,1,158,0,1.6,1,0,2,1\r\n58,0,2,120,340,0,1,172,0,0,2,0,2,1\r\n66,0,3,150,226,0,1,114,0,2.6,0,0,2,1\r\n43,1,0,150,247,0,1,171,0,1.5,2,0,2,1\r\n69,0,3,140,239,0,1,151,0,1.8,2,2,2,1\r\n59,1,0,135,234,0,1,161,0,0.5,1,0,3,1\r\n44,1,2,130,233,0,1,179,1,0.4,2,0,2,1\r\n42,1,0,140,226,0,1,178,0,0,2,

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv('heart.csv')

In [ ]:
df.columns

Index(['age', 'sex', 'cp', 'trtbps', 'chol', 'fbs', 'restecg', 'thalachh',
       'exng', 'oldpeak', 'slp', 'caa', 'thall', 'output'],
      dtype='object')

### Penjelasan fitur yang ada

Age : Age of the patient

Sex : Sex of the patient

cp : Chest Pain type chest pain type

- Value 1: typical angina
- Value 2: atypical angina
- Value 3: non-anginal pain
- Value 4: asymptomatic

trtbps : resting blood pressure (in mm Hg)

chol : cholestoral in mg/dl fetched via BMI sensor

fbs : (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)

rest_ecg : resting electrocardiographic results

- Value 0: normal
- Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
- Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria

thalach : maximum heart rate achieved

exang : exercise induced angina (1 = yes; 0 = no)

oldpeak : ST depression induced by exercise relative to rest

ca : number of major vessels (0-3)

slp : the slope of the peak exercise ST segment

- Value 1: upsloping
- Value 2: flat
- Value 3: downsloping

caa : number of major vessels (0-3) colored by flourosopy

thall : (?)

output (target) : 0 = less chance of heart attack 1= more chance of heart attack

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trtbps    303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalachh  303 non-null    int64  
 8   exng      303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slp       303 non-null    int64  
 11  caa       303 non-null    int64  
 12  thall     303 non-null    int64  
 13  output    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [ ]:
df.describe()

age         sex          cp      trtbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg    thalachh        exng     oldpeak         slp         caa  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

            thall      output  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [ ]:
df.isnull().sum()

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
oldpeak     0
slp         0
caa         0
thall       0
output      0
dtype: int64

In [ ]:
# Lakukan missing value handling dengan menggunakan imputasi maupun dropping

In [ ]:
# Lengkapi list berikut

categorical_col = ["sex", "cp", "fbs", "restecg", "exng", "slp", "caa", "thall"]
numeric_col = ["age", "trtbps", "chol", "thalachh", "oldpeak" ]

In [ ]:
df[categorical_col].nunique()

sex        2
cp         4
fbs        2
restecg    3
exng       2
slp        3
caa        5
thall      4
dtype: int64

In [ ]:
# Lakukan encoding dari data kategorikal
from sklearn.preprocessing import OneHotEncoder

In [ ]:
# Lakukan tambahan preprocessing (Scaling, Normalization, Imbalance Handling, dsb) jika dirasa diperlukan (Opsional)
from sklearn.preprocessing import MinMaxScaler, normalize
scaler = MinMaxScaler()
scaled = scaler.fit_transform(df[numeric_col])
df.loc[:, numeric_col] = scaled

<ipython-input-41-bbe9995a688d>:5: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[0.70833333 0.16666667 0.25       0.5625     0.58333333 0.58333333
 0.5625     0.3125     0.47916667 0.58333333 0.52083333 0.39583333
 0.41666667 0.72916667 0.60416667 0.4375     0.60416667 0.77083333
 0.29166667 0.83333333 0.625      0.3125     0.27083333 0.66666667
 0.22916667 0.875      0.625      0.45833333 0.75       0.5
 0.25       0.75       0.3125     0.52083333 0.45833333 0.35416667
 0.52083333 0.52083333 0.75       0.75       0.45833333 0.39583333
 0.33333333 0.5        0.20833333 0.47916667 0.3125     0.375
 0.5        0.5        0.45833333 0.77083333 0.6875     0.3125
 0.70833333 0.47916667 0.39583333 0.33333333 0.10416667 0.58333333
 0.875      0.52083333 0.47916667 0.25       0.60416667 0.125
 0.45833333 0.33333333 0.3125     0.6875     0.52083333 0.45833333
 0.         0.45833333 0.29166667 0.54166667 0.45833333 0.625
 0

In [ ]:
from sklearn.model_selection import train_test_split

X = df.drop('output', axis=1).to_numpy()
y = df['output'].to_numpy()

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Implementasi Backpropagation

In [ ]:
import numpy as np
import time


# Referensi Implementasi
class BPNN:
    def __init__(self, layer_conf, learn_rate=0.1, max_epoch=10000, max_error=0.1, print_per_epoch=100):
        self.layer_conf = layer_conf
        self.learn_rate = learn_rate
        self.max_epoch = max_epoch
        self.max_error = max_error
        self.print_per_epoch = print_per_epoch
        self.weights = None

    @staticmethod
    def _sig(x):
        return 1 / (1 + np.exp(-x))

    @staticmethod
    def _sigd(x):
        s = BPNN._sig(x)
        return s * (1 - s)

    def fit(self, X, target):
        start_time = time.time()
        np.random.seed(1)

        nin = [np.empty(i) for i in self.layer_conf]
        n = [np.empty(j + 1) if i < len(self.layer_conf) - 1 else np.empty(j)
             for i, j in enumerate(self.layer_conf)]
        self.weights = [np.random.rand(self.layer_conf[i] + 1, self.layer_conf[i + 1])
                for i in range(len(self.layer_conf) - 1)]
        dw = [np.empty((self.layer_conf[i] + 1, self.layer_conf[i + 1])) for i in range(len(self.layer_conf) - 1)]
        d = [np.empty(s) for s in self.layer_conf[1:]]
        din = [np.empty(s) for s in self.layer_conf[1:-1]]

        epoch = 0
        mse = 1

        for i in range(0, len(n) - 1):
            n[i][-1] = 1

        while (self.max_epoch == -1 or epoch < self.max_epoch) and mse > self.max_error:
            epoch += 1
            mse = 0

            for r in range(len(X)):
                n[0][:-1] = X[r]

                for L in range(1, len(self.layer_conf)): # Forward Pass
                    nin[L] = np.dot(n[L - 1], self.weights[L - 1])
                    n[L][:len(nin[L])] = self._sig(nin[L])

                e = target[r] - n[-1] # MSE
                mse += sum(e ** 2)

                d[-1] = e * self._sigd(nin[-1]) # Backprop
                dw[-1] = self.learn_rate * d[-1] * n[-2].reshape((-1, 1))

                for L in range(len(self.layer_conf) - 1, 1, -1):
                    din[L - 2] = np.dot(d[L - 1], np.transpose(self.weights[L - 1][:-1]))
                    d[L - 2] = din[L - 2] * np.array(self._sigd(nin[L - 1]))
                    dw[L - 2] = (self.learn_rate * d[L - 2]) * n[L - 2].reshape((-1, 1))

            for L in range(len(self.weights)): # Update Bobot GD
                    self.weights[L] += dw[L]

            mse /= len(X)

            if self.print_per_epoch > -1 and epoch % self.print_per_epoch == 0:
                print(f'Epoch {epoch}, MSE: {mse}')

        execution = time.time() - start_time
        print("Execution Time: {:.2f} seconds".format(execution))
        return self

    def predict(self, X):
        n = [np.empty(len(i)) for i in self.weights]
        nin = [np.empty(len(i[0])) for i in self.weights]
        predict = []
        n.append(np.empty(len(self.weights[-1][0])))

        for x in X:
            n[0][:-1] = x
            for L in range(0, len(self.weights)):
                nin[L] = np.dot(n[L], self.weights[L])
                n[L + 1][:len(nin[L])] = self._sig(nin[L])
            predict.append(n[-1].copy())

        return predict


In [ ]:
class BPNN_SGD(BPNN):
    def fit(self, X, target):
        start_time = time.time()
        np.random.seed(1)

        nin = [np.empty(i) for i in self.layer_conf]
        n = [np.empty(j + 1) if i < len(self.layer_conf) - 1 else np.empty(j)
             for i, j in enumerate(self.layer_conf)]
        self.weights = [np.random.rand(self.layer_conf[i] + 1, self.layer_conf[i + 1])
                for i in range(len(self.layer_conf) - 1)]
        dw = [np.empty((self.layer_conf[i] + 1, self.layer_conf[i + 1])) for i in range(len(self.layer_conf) - 1)]
        d = [np.empty(s) for s in self.layer_conf[1:]]
        din = [np.empty(s) for s in self.layer_conf[1:-1]]

        epoch = 0
        mse = 1

        for i in range(0, len(n) - 1):
            n[i][-1] = 1

        while (self.max_epoch == -1 or epoch < self.max_epoch) and mse > self.max_error:
            epoch += 1
            mse = 0

            for r in range(len(X)):
                n[0][:-1] = X[r]

                for L in range(1, len(self.layer_conf)):  # Forward Pass
                    nin[L] = np.dot(n[L - 1], self.weights[L - 1])
                    n[L][:len(nin[L])] = self._sig(nin[L])

                e = target[r] - n[-1]  # MSE
                mse += sum(e ** 2)

                d[-1] = e * self._sigd(nin[-1])  # Backprop
                dw[-1] = self.learn_rate * d[-1] * n[-2].reshape((-1, 1))

                for L in range(len(self.layer_conf) - 1, 1, -1):
                    din[L - 2] = np.dot(d[L - 1], np.transpose(self.weights[L - 1][:-1]))
                    d[L - 2] = din[L - 2] * np.array(self._sigd(nin[L - 1]))
                    dw[L - 2] = (self.learn_rate * d[L - 2]) * n[L - 2].reshape((-1, 1))

                for L in range(len(self.weights)):  # Update Bobot SGD
                    self.weights[L] += dw[L]

            mse /= len(X)

            if self.print_per_epoch > -1 and epoch % self.print_per_epoch == 0:
                print(f'Epoch {epoch}, MSE: {mse}')

        execution = time.time() - start_time
        print("Execution Time: {:.2f} seconds".format(execution))
        return self


In [ ]:
from sklearn.metrics import f1_score, confusion_matrix

gd = BPNN(layer_conf=[13, 5, 1], learn_rate=0.1, max_epoch=1000, max_error=0.01, print_per_epoch=100)
gd.fit(X_train, y_train)
y_pred_gd = np.round(gd.predict(X_test)).astype(int)


Epoch 100, MSE: 0.398982649107622
Epoch 200, MSE: 0.40634282064754557
Epoch 300, MSE: 0.41126475285279473
Epoch 400, MSE: 0.4148478341390693
Epoch 500, MSE: 0.4176043517992363
Epoch 600, MSE: 0.4198089511210715
Epoch 700, MSE: 0.4216236750009628
Epoch 800, MSE: 0.42315104887626176
Epoch 900, MSE: 0.4244594946737089
Epoch 1000, MSE: 0.4255966381544647
Execution Time: 22.33 seconds


In [ ]:
sgd = BPNN_SGD(layer_conf=[13, 5, 1], learn_rate=0.1, max_epoch=1000, max_error=0.01, print_per_epoch=100)
sgd.fit(X_train, y_train)
y_pred_sgd = np.round(sgd.predict(X_test)).astype(int)

Epoch 100, MSE: 0.11206989718455518
Epoch 200, MSE: 0.10611747368405125
Epoch 300, MSE: 0.09998743734525993
Epoch 400, MSE: 0.08343082566413185
Epoch 500, MSE: 0.07612533198902253
Epoch 600, MSE: 0.07372324800826775
Epoch 700, MSE: 0.07223899164739776
Epoch 800, MSE: 0.07091120161435432
Epoch 900, MSE: 0.0695635735133294
Epoch 1000, MSE: 0.06816123363371086
Execution Time: 18.64 seconds


In [51]:
# Gunakan data yang telah dipersiapkan sebelumnya untuk melatih model Backpropagation dengan SGD dan lakukan evaluasi terhadap akurasi
# Ujikan kedua model GD dan SGD untuk kemudian dibandingkan
# Jika dirasa perlu, model dapat dilanjutkan dievaluasi menggunakan F1-Score dan Confusion Matrix (Opsional)

from sklearn.metrics import f1_score, confusion_matrix, accuracy_score

accuracy_gd = accuracy_score(y_test, y_pred_gd)
f1_gd = f1_score(y_test, y_pred_gd)
cm_gd = confusion_matrix(y_test, y_pred_gd)
print(f"Akurasi GD: {accuracy_gd:.2f}")
print(f"F1-Score GD: {f1_gd:.2f}")
print("Confusion Matrix GD:\n", cm_gd)

accuracy_sgd = accuracy_score(y_test, y_pred_sgd)
f1_sgd = f1_score(y_test, y_pred_sgd)
cm_sgd = confusion_matrix(y_test, y_pred_sgd)
print(f"Akurasi SGD: {accuracy_sgd:.2f}")
print(f"F1-Score SGD: {f1_sgd:.2f}")
print("Confusion Matrix SGD:\n", cm_sgd)

Akurasi GD: 0.52
F1-Score GD: 0.69
Confusion Matrix GD:
 [[ 0 29]
 [ 0 32]]
Akurasi SGD: 0.77
F1-Score SGD: 0.75
Confusion Matrix SGD:
 [[26  3]
 [11 21]]


## Analisis


1. Apabila menggunakan preprocessing data tambahan, jelaskan mengapa anda melakukan preprocessing tersebut!
2. Apa perbedaan dari Gradient Descent dan Stochastic Gradient Descent dan apa kelebihan dari SGD?
3. Jika model dievaluasi menggunakan F1-Score dan Confusion Matrix jelaskan hasil evaluasi yang anda dapatkan!
4. Apakah Penggunaan SGD dapat membantu model untuk mendapatkan akurasi yang lebih baik?

## Jawaban

1. Untuk menyamakan rentang dari data saya menggunakan MinMax. karena dataset rentangnya terlihat sangat signifikan

2. Perbedaannya untuk GD menghitung gradien untuk sleuruh dataset sebelum memperbarui bobot sedangksn SGD memperbarui bobot berdasarkan gradien dari mini batch tiap iterasi.

3. Untuk F!, SGD lebih tinggi karena precision dan recallnya konsisten, untuk Confusion matrix  GD tidak ada prediksi true positif untuk kelas negatif dan untuk Confusion matrix SGD prediksinya seimbang dengan 26 TN, 3 FP, 11 FN, dan 21 TP

4. tentu saja, penggunaan SGD meningkatkan akurasi(dilihat dari hasil evaluasi)